### PR TITLE
[ADD] .github: introduce CODEOWNERS

### PR DESCRIPTION
**Pull request by seb-odoo**
Code owners are automatically requested for review when someone opens a pull
request that modifies code that they own.

Code owners are not automatically requested to review draft pull requests.

https://help.github.com/en/github/creating-cloning-and-archiving-repositories/about-code-owners

---------------

It is optional. If you want to participate, please update the list in this PR according to your preferences. If your team does not exist yet on github, ask a repo admin to create it.

### CODE REVIEW

**Comment by seb-odoo** [comment]
`/addons/account/tests/test_portal_attachment.py @seb-odoo`

**Comment by mart-e** [comment]
> # Prefer using teams whenever possible, as members of teams can be dynamically
> # managed outside of this file without requiring a new commit.

@seb-odoo why are you code owner of a shitload of files? This is typically why I am not a fan of codeowner files, mixing organisation with code.

**Comment by seb-odoo** [comment]
I would tend to agree with you, this is still open to discussion.

The goal of this first draft is to see where everyone would want to go with it, and then we can decide what patterns are actually reasonable to keep.

Do you have a suggestion that would solve your concern?

**Comment by mart-e** [comment]
> Do you have a suggestion that would solve your concern?

If we want to use codeowner files, we should only use teams I think (even if containing only one guy). What I am afraid is that we will need to make a lot of changes all the time to keep it consistent: we are adding/renaming/deleting modules like all the time. We are also changing team scope all the time (apps like sale/crm/hr changing of responsability often).

On the other hand, it's cool if it allow us to review more contribution from other teams/exterior contributors.

Why not go the Odoo's way and make it ourself in a module?

File: .github/CODEOWNERS [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+
+/odoo/addons/base/tests/test_image.py @seb-odoo
+/odoo/addons/base/models/ir_attachment.py @seb-odoo
+
+/odoo/http.py @odoo/rd-security
+/odoo/sql_db.py @odoo/rd-security
+
+# expression, query
+/odoo/osv/** @odoo/rd-security
+
```
Comment (KangOl) [inline .github/CODEOWNERS]:
```suggestion

/odoo/modules/migration.py @odoo/upgrade
```

Comment (seb-odoo) [reply .github/CODEOWNERS]: Added

**Comment by seb-odoo** [comment]
Removed all occurrences of users in favor of groups:

- created teams: @odoo/rd-images and @odoo/rd-code-owners
- added @odoo/rd-discuss instead of myself for `test_mail*` stuff, even though it is not exactly the same scope
- removed myself from product stuff, I didn't feel it was worth creating a dedicated group

Let's start with this and see how it goes in practice.

@odony please make the final check, and if you approve my new teams, feel free to merge & give the proper write permissions to the teams mentioned above (and @odoo/upgrade if it doesn't have it yet).

**Comment by seb-odoo** [comment]
...and now with order guidelines properly followed.

**Comment by JKE-be** [comment]
Seems like you are mixin what is you interest in Odoo and the notion of code owner.

I don't see the sense to have rd-image = seb-odoo reviewer of something that I add in tools or qweb part.

If you continue to add you as codeowner on each file that you touch or read one day, it remove all the sense, except to have your name on each PR as reviewer for decoration.

**Comment by JKE-be** [comment]
> Why not go the Odoo's way and make it ourself in a module?

+1, Website team will be happy to know that some change are done in a file that touch website or http e.g. 
But if we are reviewer of each PR, we loose the focus on PR that really need review by website team.

**Comment by seb-odoo** [comment]
I think it's just a matter of time and priority, Github already has the feature somewhat implemented so we want to test it while we don't have a module ready?

But otherwise I agree, I would prefer having a module sending notifications too since it would be more flexible for us compared to what Github decides to make.

The reasoning for the review is just we can remove ourselves from review if we are not interested in a particular PR but at least we get the notification.

--

In my particular case it's not just files where I'm interested, I will review the PR that I get this way, and if I get too much of them, I will restrict the patterns accordingly.

I'm actually more afraid (of spam) of all the generic patterns that legitimately concern "discuss" than all of the patters for "image" that are targeting files rarely modified.

For `tools` considering it rarely changes it is easier to have a simple and generic pattern, as opposed to creating separately one for `image` file, one for `mimetype`, etc. I don't mind to register those if we think it is better that way.

For `ir_qweb` this is the one from `web` not the generic one. There is only the `Image` class on it.

**Comment by JKE-be** [comment]
I just checked the new pricing...
It seems to be a feature for team’s plan and up only, are you sure that we can use it in our plan?
It will be not the first feature that we cannot use because we have an old plan...

**Comment by KangOl** [comment]
> I just checked the new pricing...
> It seems to be a feature for team’s plan and up only, are you sure that we can use it in our plan?
> It will be not the first feature that we cannot use because we have an old plan...

Yes, we can use for odoo/odoo repository as it is public. However, we don't have this option for our private repositories (odoo/enterprise).

File: .github/CODEOWNERS [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+# ===============
+
+# Prefer using teams whenever possible, as members of teams can be dynamically
+# managed outside of this file without requiring a new commit.
+# Only define users for specific cases.
+# For each pattern: list teams first, then users, both alphabetically.
+
+# Keep in mind the following requirements for the feature to work:
+# - Listed teams and users must have write permissions on the repo.
+# - For teams "code review assignment" has to be properly configured in the
```
[the remark this answers is not in the document]
Comment (seb-odoo) [reply .github/CODEOWNERS:28]: This is a great find, I will adapt accordingly.

File: .github/CODEOWNERS [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+# When creating a more specific pattern of an existing more generic pattern:
+# - The specific pattern must be listed below.
+# - Teams and users of the generic pattern have to be copied to the specific
+#   pattern, unless they approve to lose their ownership for the paths matched
+#   by the specific pattern.
+
+# Teams and users
+# ===============
+
+# Prefer using teams whenever possible, as members of teams can be dynamically
```
[the remark this answers is not in the document]
Comment (seb-odoo) [reply .github/CODEOWNERS:21]: ok

File: .github/CODEOWNERS [hunk 4]
```diff
@@ -0,0 +1,65 @@
+# Doc: https://help.github.com/en/github/creating-cloning-and-archiving-repositories/about-code-owners
+# Order is important; the last matching pattern takes the most precedence.
+
+# Please apply the following guidelines when updating this file.
+
+# Paths and patterns
+# ==================
+
+# Path should be ordered alphabetically (when precedence is not an issue).
+# Avoid complex patterns, prefer full path or with wildcard only at the end.
```
[the remark this answers is not in the document]
Comment (seb-odoo) [reply .github/CODEOWNERS:10]: I suppose no useless `*` is better, I will make changes and retest some of them.

**Comment by seb-odoo** [comment]
Changes done.

**Comment by odony** [comment]
Also verified that all 7 teams currently mentioned have write access on the odoo/odoo repo, and have review assignation disabled (i.e. only assign the team globally)

**Comment by tde-banana-odoo** [comment]
Side question: why not in stable, at least 13 ? Notably for mail fixes concerning me, myself and I . Tant pis ?

**Comment by odony** [comment]
> Side question: why not in stable, at least 13 ? Notably for mail fixes concerning me, myself and I . Tant pis ?

It's mostly a PoC at first, we can start doing this in stable if it works well in master.

**Comment by tde-banana-odoo** [comment]
Ok, seems fair. As first notifications are pushed I suppose there will be some update soon :) .

**Comment by seb-odoo** [comment]
Quick feedback after 1 day:

More notifications that expected, but this is in big part due to catching up on existing PR that also get tagged after any update on them.

A lot of notifications received "too early" (for my taste) due to PR not being created as draft. While it can be good to catch some stuff early, I feel like getting notifications for all updates that will happen in those PR in the future will be too much, and I'm afraid if I unsubscribe now I won't see it when it is actually ready.
I wonder if I can set those PR as draft, unsubscribe, and get the notification again when they become ready.
Let's give it a few more days, and maybe we need to educate the teams about the usage of draft.

Some rules might need to be more specific (first example in mind: those in /tools) and I'm not sure why we set `rd-discuss` to follow `test_mail_full` and the other variants instead of `rd-sm`. Those are relatively minor and can be tweaked easily.

Now the good point, except those cases mentioned above, overall I like the notifications I received so far, most of them were appropriate, and some already allowed to catch few issues (relatively minor so far).

**Pull request closed without merging**